### PR TITLE
Begin stableswap spec

### PR DESCRIPTION
**Pull request by ValarDragon**
## What is the purpose of the change

WIP work on a spec for Solidly stableswap and description of equations used within it

Rendered URL: https://github.com/osmosis-labs/osmosis/blob/dev/stableswap_spec/x/gamm/pool-models/stableswap/README.md

### CODE REVIEW

File: x/gamm/pool-models/stableswap/README.md [hunk 1]
```diff
@@ -1,4 +1,106 @@
...
+
+<!---  This is true, but however the details of this curve actually lend itself better to using more general solutions for swaps via binary search. (Inspired from https://arxiv.org/abs/2111.13740 ) The remainder of the methods we implement via generic CFMM ideas.
+
+We detail below both what the direct solution via the CFMM equation is, and the solution we use.
+Then we show how these are used to give all of the swqp equations. --->
+
+### CFMM function
+
+Most operations we do only need to reason about two of the assets in a pool, and sometimes only one.
+We wish to have a simpler CFMM function to work with in these cases.
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/README.md:45]:
```suggestion
We wish to have a simpler CFMM function to work within these cases.
```

File: x/gamm/pool-models/stableswap/README.md [hunk 2]
```diff
@@ -1,4 +1,106 @@
...
+Due to the CFMM equation $f$ being a symmetric function, we can wlog reorder the arguments to the function. Thus we put the assets of relevance at the beginning of the function. So if two assets $x, y$, we write: $f(x,y, a_3, ... a_n) = xy * a_3 * ... a_n (x^2 + y^2 + a_3^2 + ... + a_n^2)$.
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/README.md:46]: nit: I didn't realize right away what this abbreviation means and had to search
```suggestion
Due to the CFMM equation $f$ being a symmetric function, we can without loss of generality reorder the arguments to the function. Thus we put the assets of relevance at the beginning of the function. So if two assets $x, y$, we write: $f(x,y, a_3, ... a_n) = xy * a_3 * ... a_n (x^2 + y^2 + a_3^2 + ... + a_n^2)$.
```

File: x/gamm/pool-models/stableswap/README.md [hunk 3]
```diff
@@ -1,4 +1,106 @@
...
+
+We then take a more convenient expression to work with, via variable substition.
+$$ \begin{equation}
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/README.md:49]: This isn't rendering for me

File: x/gamm/pool-models/stableswap/README.md [hunk 4]
```diff
@@ -1,4 +1,106 @@
...
+First notice that for all swaps, we only deal with two assets at a time, as swaps are given one asset in, and one asset out. For exposition, lets call the input asset $x$, the output asset $y$, and we can compute $v$ and $w$ given the other assets, whose reserves are untouched.
+
+First we note the direct way of solving this, its limitation, and then the binary search equations.
+
+#### Direct swap solution
+
+Suppose the existence of a function $\text{solve\_cfmm}(x, v, w, k) = y\text{ s.t. }g(x, y, v, w) = k$.
+Then we can solve swaps by first computing $k = g(x_0, y_0, v, w)$.
+Then suppose I want to swap $a$ units of $x$, and then want to find how many units $b$ of $y$ that we get out.
+We do this by computing $y_f = \text{solve\_cfmm}(x_0 + a, v, w, k)$, and then $b = y_0 - y_f$
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/README.md:78]: This and all of the remaining expressions also do not format for me unfortunately

**Comment by AlpinYukseloglu** [comment]
Thoughts on throwing this into a hackmd so we can work on filling out the rest together? Or would you prefer to fill out the whole spec and then have others review?

File: x/gamm/pool-models/stableswap/README.md (hunk 3 above)
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/README.md:49]: Hrmm, this is official mathjax

I think we just link to something else for rendering then, I don't want to go through the trouble of replacing these with images. Or we can fix rendering in a second pass later. It should work in your IDE + hackmd

File: x/gamm/pool-models/stableswap/README.md [hunk 5]
```diff
@@ -1,4 +1,137 @@
...
+
+def binary_search(x_f, y_0, w, k, err_tolerance):
+  iter_count = 0
+  y_est = y_0
+  k_est = h(x_f, y_0, w)
+  lowerbound = 0
+  upperbound = 
+```
+
+### Spot Price
```
Comment (alexanderbez) [inline x/gamm/pool-models/stableswap/README.md:121]: Are these considered as TODOs for future reference?

Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/README.md:139]: Yeah, just things that should be filled in before stableswap has a full spec

**Comment by ValarDragon** [comment]
@p0mvn github rendering now fixed

**Comment by ValarDragon** [comment]
@AlpinYukseloglu happy with either of the options you proposed.

Feel free to commit to the branch, or move it to a hackmd. I also don't think we need to have the full spec written in one PR though. (e.g. I'm happy with this getting in its current form, I feel like it gives us a swap algorithm description we can iterate on for comparing against code)

File: x/gamm/pool-models/stableswap/README.md [hunk 6]
```diff
@@ -1,4 +1,185 @@
...
+
+Instead of using the direct solution for $\text{solve cfmm}(x, w, k')$, instead notice that $h(x, y, w)$ is an increasing function in $y$. 
+So we can simply binary search for $y$ such that $h(x, y, w) = k'$, and we are guaranteed convergence within some error bound. 
+
+In order to do a binary search, we need bounds on $y$. 
+The lowest lowerbound is $0$, and the largest upperbound is $\infty$. 
+The maximal upperbound is obviously unworkable, and in general binary searching around wide ranges is unfortunate, as we expect most trades to be centered around $y_0$. 
+This would suggest that we should do something smarter to iteratively approach the right value for the upperbound at least. 
+Notice that $h$ is super-linearly related in $y$, and at most cubically related to $y$. 
+This means that $ \forall c \in \R^+, c * h(x,y,w) < h(x,c*y,w) < c^3 * h(x,y,w)$.
```
Comment (p0mvn) [inline x/gamm/pool-models/stableswap/README.md:106]: This one isn't rendering still

**Comment by ValarDragon** [comment]
ty. fixed! (Also apologies, last commit got cut off a bit early in the web editor)

**Review by p0mvn — APPROVED** [review]
The progress so far LGTM and everything makes sense. Nice work

File: x/gamm/pool-models/stableswap/README.md [hunk 7]
```diff
@@ -1,4 +1,197 @@
...
+The method to compute this under 0 swap fee is implied by the CFMM equation itself, since the constant refers to:
+$g(x_0, y_0, v, w) = k = g(x_0 + a, y_0 - b, v, w)$. As $k$ is linearly related to $v$, and $v$ is unchanged throughout the swap, we can simplify the equation to be reasoning about $k' = \frac{k}{v}$ as the constant, and $h$ instead of $g$
+
+We then model the solution by finding a function $\text{solve cfmm}(x, w, k') = y\text{ s.t. }h(x, y, w) = k'$.
+Then we can solve the swap amount out by first computing $k'$ as $k' = h(x_0, y_0, w)$, and 
+computing $y_f := \text{solve cfmm}(x_0 + a, w, k')$. We then get that $b = y_0 - y_f$.
+
+So all we need is an equation for $\text{solve cfmm}$! Its essentially inverting a multi-variate polynomial, and in this case is solvable: [wolfram alpha link](https://www.wolframalpha.com/input?i=solve+for+y+in+x+*+y+*+%28x%5E2+%2B+y%5E2+%2B+w%29+%3D+k)
+
+Or if were clever with simplification in the two asset case, we can reduce it to: [desmos link](https://www.desmos.com/calculator/ktdvu7tdxv).
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/README.md]: ty, updated!

**Pull request merged**